File: kotlin-example/src/main/kotlin/ex06/Example06.java
package ex06;

interface Bark {
    void bark();
}

abstract class Animal implements Bark {

    private final String name;
    private int age;


    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public Animal(String name){
        this.name = name;
    }

    public void eat(){
        System.out.println(name+" 식사 시작 합니다.");
    }

    public String getName() {
        return name;
    }
}

class Dog extends Animal{

    public Dog(){
        super("");
    }

    public Dog(String name){
        super(name);
    }

    @Override
    public void bark() {
        System.out.println(this.getName()+" : 멍멍");
    }
}


public class Example06 {

    public Example06(){
        var dog = new Dog("퍼피");
        dog.setAge(10);
        var age = dog.getAge();
        dog.eat();
        dog.bark();
    }

    public static void main(String[] args){
        new Example06();
    }
}

File: kotlin-example/src/main/kotlin/ex06/Example06.kt
package org.example.ex06

fun main() {
    // new keyword 필요없음
    var dog = Dog("해피")
}


interface Bark {
    fun bark()
}

interface Temp {
    fun hi()
}

abstract class Animal (
    private val name: String? = ""
) : Bark // implements 가 여기로 옴, 상속받는 extends, implements가 여기에 따로 구분없이 쓰임
{
    fun eat() {
        println("$name 식사 시작 합니다.")
    }
}


class Dog (
    private val name: String? = null
) : Animal(name) , Temp // Dog 의 파라메터를 Animal 로 넘김 (super 역할)
{
    var age: Int? = 0
        get() = field
        set(value) {
           field = value?.let {it + 1} ?: value
        }

    override fun bark() {
        println("$name 멍멍")
    }

    override fun hi() {}
}
